pyramid/csrf.py

```diff
@@ -0,0 +1,291 @@
+from functools import partial
+import uuid
+
+from zope.interface import implementer
+
+from pyramid.compat import (
+    urlparse,
+    bytes_
+)
+from pyramid.exceptions import (
+    BadCSRFOrigin,
+    BadCSRFToken,
+)
+from pyramid.interfaces import ICSRFStoragePolicy
+from pyramid.settings import aslist
+from pyramid.util import (
+    is_same_domain,
+    strings_differ
+)
+
+
+@implementer(ICSRFStoragePolicy)
+class SessionCSRF(object):
+    """ The default CSRF implementation, which mimics the behavior from older
+    versions of Pyramid. The ``new_csrf_token`` and ``get_csrf_token`` methods
+    are indirected to the underlying session implementation.
+
+    Note that using this CSRF implementation requires that
+    a :term:`session factory` is configured.
+
+    .. versionadded :: 1.8a1
+    """
+    def new_csrf_token(self, request):
+        """ Sets a new CSRF token into the session and returns it. """
+        return request.session.new_csrf_token()
+
+    def get_csrf_token(self, request):
+        """ Returns the currently active CSRF token from the session, generating
+        a new one if needed."""
+        return request.session.get_csrf_token()
+
+    def check_csrf_token(self, request, supplied_token):
+        """ Returns True if supplied_token is the same value as get_csrf_token
+        returns for this request. """
+        expected = self.get_csrf_token(request)
+        return not strings_differ(
+            bytes_(expected, 'ascii'),
+            bytes_(supplied_token, 'ascii'),
+        )
+
+@implementer(ICSRFStoragePolicy)
+class CookieCSRF(object):
+    """ An alternative CSRF implementation that stores its information in
+    unauthenticated cookies, known as the 'Double Submit Cookie' method in the
+    OWASP CSRF guidelines. This gives some additional flexibility with regards
+    to scaling as the tokens can be generated and verified by a front-end
+    server.
+
+    .. versionadded :: 1.8a1
+    """
+
+    def __init__(self, cookie_name='csrf_token', secure=False, httponly=False,
+                 domain=None, path='/'):
+        self.cookie_name = cookie_name
+        self.secure = secure
+        self.httponly = httponly
+        self.domain = domain
+        self.path = path
```

Comment:
This implementation should be using the `webob.cookies.CookieProfile` from webob in order to set the cookies on potentially multiple domains similar to how `SignedCookieSessionFactory` and `AuthTktAuthenticationPolicy` operate.